examples/merkle_proofs.md

```diff
@@ -0,0 +1,157 @@
+
+```ocaml
+# let tree = Lwt_main.run (
+    let tree = Store.Tree.empty () in
+    let* tree = Store.Tree.add tree [ "b"; "e"; "n" ] 10 in
+    let* tree = Store.Tree.add tree [ "b"; "o"; "b" ] 20 in
+    let+ tree = Store.Tree.add tree [ "e"; "v"; "e" ] 30 in
+    tree
+  );;
+val tree : Store.tree = <abstr>
+```
+
+In order to produce a Merkle proofs, Irmin requires that the tree on which the proof will be based is persisted inside a `repo`. The following code block does so.
+
+`tree_key` is a value that encodes where `tree` has been persisted inside the store's backend.
+
+```ocaml
+# let tree_key = Lwt_main.run (
+    (* [batch] exposes [repo] stores in read-write mode *)
+    let+ kinded_key = Store.Backend.Repo.batch repo
+      (fun rw_contents_store rw_node_store _rw_commit_store ->
+        Store.save_tree repo rw_contents_store rw_node_store tree)
+    in
+    (* As the root of [tree] is a node, refining the type is safe *)
+    match kinded_key with
+    | `Node key -> key
+    | `Contents _ -> assert false
+  );;
+val tree_key : Store.node_key = <abstr>
+```
+
+The bank is ready. `repo` contains the data, which was persisted on disk. `tree_key` acts as a pointer to the root of the accounts tree.
+
+Since this example only cares about account statements, there is no need to worry about modifying the account balances, this would be well expressed using Irmin's commit and branches abstraction, but this is out of the scope of this example.
+
+![The bank's data](./merkle_proof_1.jpg)
+
+### Producing Acccount Statements
+
+Let's produce an account statement for Eve.
+
+```ocaml
+let visit_tree tree =
+  (* [tree] is shallow. Let's only load the parts we are interested in *)
+  let+ (_ : int option) = Store.Tree.find tree [ "e"; "v"; "e" ] in
+  (Store.Tree.empty (), `Success)
+
+let proof, `Success = Lwt_main.run (
+  Store.Tree.produce_proof repo (`Node tree_key) visit_tree)
+
+let pp_merkle_proof = Irmin.Type.pp Store.Tree.Proof.tree_t
+```
+
+```ocaml
+# let () = Fmt.pr "proof: %a\n" pp_merkle_proof (Store.Tree.Proof.state proof);;
+proof: {"Node":[["b",{"Blinded_node":"584ba234368730ce6cea56d93433a7a777cad2a0"}],["e",{"Node":[["v",{"Node":[["e",{"Contents":[30,"normal"]}]]}]]}]]}
+```
+
+The proof of Eve's balance is a Merkle proof. It is a subset of the tree of all the accounts in the bank, it doesn't leaks Bob and Ben's balance, it only reveals the hash and the name of their subtree (i.e. `["b",{"Blinded_node":"584ba234368730ce6cea56d93433a7a777cad2a0"}]`).
+
+![The Merkle proof](./merkle_proof_2.jpg)
+
+The reason why this statement is temper proof is because it hashes the same way as the full account tree of the bank. The bank could make public the hash in a trusted place (e.g. on Twitter) and anyone could produce the hash of the statement and make sure that it is identical to the hash published by the bank.
```

Comment:
"The bank could make public the hash in an untrusted place (e.g. on Twitter)"?
Nevertheless, I'd need more reflection on this paragraph in order to figure out whether it is really the case that "this statement is temper proof". Namely, there are attacks which produce a (different) tree compatible with the public hash. Is the implementation containing mitigations for this type of attacks?

Reply:
> Namely, there are attacks which produce a (different) tree compatible with the public hash. Is the implementation containing mitigations for this type of attacks?

Hash collision attacks of this sort would be a security vulnerability in Irmin, which is quite careful about ensuring that different trees cannot have the same hash. AFAIK, we're not mitigating against this outcome because if it were to happen the trust relationship between all parties is already untenable.

Reply:
"temper" should be "tamper"

Reply:
Good point Tom!
Regarding the hash collision, could we have more details about the guarantees provided by Irmin that different trees cannot have the same hash? Or what's the meaning of "quite careful"?